File: web/src/lib/removedAnswerKeys.ts
/** Copier answer keys removed from the public template surface. */

/** Keep in parity with `src/riso/core/removed_answer_keys.py`. */
export const REMOVED_ANSWER_KEYS: Record<string, string> = {
  api_tracks: '`api_module` plus `api_languages`',
  api_language: '`api_languages`',
  docs_site: '`docs_module` plus `docs_framework`',
  mcp_language: '`mcp_languages`',
  saas_starter_module: '`saas_infra_module`',
  saas_auth: '`saas_auth_module` plus `saas_auth_provider`',
  saas_billing: '`saas_billing_module` plus `saas_billing_provider`',
  include_admin: '`saas_admin_dashboard`',
}

export type RemapAction =
  | 'wrap-list'
  | 'derive'
  | 'rename'
  | 'split'
  | 'rename-bool'

export interface RemapOp {
  old: string
  new_keys: readonly string[]
  action: RemapAction
  before?: unknown
  after?: Record<string, unknown>
}

export interface RemapResult {
  answers: Record<string, unknown>
  ops: RemapOp[]
}

export const ANSWER_KEY_REMAPS: Record<string, RemapOp> = {
  api_tracks: {
    old: 'api_tracks',
    new_keys: ['api_module', 'api_languages'],
    action: 'derive',
  },
  api_language: {
    old: 'api_language',
    new_keys: ['api_languages'],
    action: 'wrap-list',
  },
  docs_site: {
    old: 'docs_site',
    new_keys: ['docs_module', 'docs_framework'],
    action: 'derive',
  },
  mcp_language: {
    old: 'mcp_language',
    new_keys: ['mcp_languages'],
    action: 'wrap-list',
  },
  saas_starter_module: {
    old: 'saas_starter_module',
    new_keys: ['saas_infra_module'],
    action: 'rename',
  },
  saas_auth: {
    old: 'saas_auth',
    new_keys: ['saas_auth_module', 'saas_auth_provider'],
    action: 'split',
  },
  saas_billing: {
    old: 'saas_billing',
    new_keys: ['saas_billing_module', 'saas_billing_provider'],
    action: 'split',
  },
  include_admin: {
    old: 'include_admin',
    new_keys: ['saas_admin_dashboard'],
    action: 'rename-bool',
  },
}

const TOKEN_SPLIT = /[\s,+/|]+/
const API_LANGUAGES = new Set(['python', 'node', 'rust', 'go'])
const MCP_LANGUAGES = new Set(['python', 'typescript', 'rust', 'go'])
const MCP_ALIASES: Record<string, string> = { node: 'typescript', js: 'typescript' }
const API_TRACK_ALIASES: Record<string, string> = {
  python: 'python',
  node: 'node',
  rust: 'rust',
  go: 'go',
  fastapi: 'python',
  fastify: 'node',
  actix: 'rust',
}
const API_TRACKS_OFF = new Set(['', 'none', 'disabled', '[]'])
const DOCS_SITE_OFF = new Set(['none', 'false', 'disabled', 'off'])
const DOCS_SITE_FRAMEWORKS: Record<string, string> = {
  sphinx: 'sphinx-shibuya',
  'sphinx-shibuya': 'sphinx-shibuya',
  docusaurus: 'docusaurus',
  fumadocs: 'fumadocs',
}
const SPLIT_OFF = new Set(['none', 'disabled', 'false', 'off'])
const SAAS_AUTH_PROVIDERS = new Set(['clerk', 'authjs'])
const SAAS_BILLING_PROVIDERS = new Set(['stripe', 'paddle', 'lemonsqueezy'])
const MODULE_ON = new Set(['enabled', 'true', 'yes', 'on', '1'])
const MODULE_OFF = new Set(['disabled', 'false', 'no', 'off', '0', 'none'])
const BOOL_TRUE = new Set(['true', 'yes', 'on', '1', 'enabled'])
const BOOL_FALSE = new Set(['false', 'no', 'off', '0', 'disabled', 'none', ''])

export class RemovedAnswerKeyError extends Error {
  readonly errors: string[]

  constructor(errors: string[]) {
    super(
      `Removed Copier answer keys are no longer supported:\n${errors
        .map((line) => `- ${line}`)
        .join('\n')}`,
    )
    this.name = 'RemovedAnswerKeyError'
    this.errors = errors
  }
}

function isNone(value: unknown): boolean {
  return value === null || value === undefined
}

function isSequence(value: unknown): value is unknown[] {
  return Array.isArray(value)
}

function normToken(value: unknown): string {
  if (isNone(value)) {
    return 'none'
  }
  return String(value).trim().toLowerCase()
}

function tokens(value: unknown): string[] {
  if (isNone(value)) {
    return []
  }
  if (isSequence(value)) {
    return value.map((item) => normToken(item)).filter((item) => item.length > 0)
  }
  const text = String(value).trim()
  if (!text) {
    return []
  }
  return text.toLowerCase().split(TOKEN_SPLIT).filter((part) => part.length > 0)
}

function wrapList(
  value: unknown,
  allowed: ReadonlySet<string>,
  aliases: Record<string, string> = {},
): string[] | null {
  if (isSequence(value)) {
    const out: string[] = []
    const seen = new Set<string>()
    for (const item of value) {
      const raw = String(item).trim()
      if (!raw) {
        continue
      }
      const mapped = aliases[raw.toLowerCase()] ?? raw.toLowerCase()
      const token = allowed.has(mapped) ? mapped : raw
      if (seen.has(token)) {
        continue
      }
      seen.add(token)
      out.push(token)
    }
    return out
  }
  if (typeof value === 'string') {
    const raw = value.trim()
    if (!raw) {
      return null
    }
    const mapped = aliases[raw.toLowerCase()] ?? raw.toLowerCase()
    if (!allowed.has(mapped)) {
      return null
    }
    return [mapped]
  }
  return null
}

function moduleToggle(value: unknown): string | null {
  if (typeof value === 'boolean') {
    return value ? 'enabled' : 'disabled'
  }
  if (typeof value === 'number' && (value === 0 || value === 1)) {
    return value === 1 ? 'enabled' : 'disabled'
  }
  const text = normToken(value)
  if (MODULE_ON.has(text)) {
    return 'enabled'
  }
  if (MODULE_OFF.has(text)) {
    return 'disabled'
  }
  return null
}

function asBool(value: unknown): boolean | null {
  if (typeof value === 'boolean') {
    return value
  }
  if (typeof value === 'number' && (value === 0 || value === 1)) {
    return Boolean(value)
  }
  if (typeof value === 'string') {
    const text = value.trim().toLowerCase()
    if (BOOL_TRUE.has(text)) {
      return true
    }
    if (BOOL_FALSE.has(text)) {
      return false
    }
  }
  return null
}

function isOff(value: unknown, offTokens: ReadonlySet<string>): boolean {
  if (isNone(value)) {
    return true
  }
  if (isSequence(value) && value.length === 0) {
    return true
  }
  if (typeof value === 'boolean') {
    return !value && offTokens.has('false')
  }
  return offTokens.has(normToken(value))
}

function mapApiLanguage(value: unknown): Record<string, unknown> | null {
  const wrapped = wrapList(value, API_LANGUAGES)
  if (wrapped === null) {
    return null
  }
  return { api_languages: wrapped }
}

function mapMcpLanguage(value: unknown): Record<string, unknown> | null {
  const wrapped = wrapList(value, MCP_LANGUAGES, MCP_ALIASES)
  if (wrapped === null) {
    return null
  }
  return { mcp_languages: wrapped }
}

function mapApiTracks(value: unknown): Record<string, unknown> | null {
  if (isOff(value, API_TRACKS_OFF)) {
    return { api_module: 'disabled' }
  }
  const langs: string[] = []
  const seen = new Set<string>()
  for (const token of tokens(value)) {
    const mapped = API_TRACK_ALIASES[token]
    if (!mapped || seen.has(mapped)) {
      continue
    }
    seen.add(mapped)
    langs.push(mapped)
  }
  if (langs.length === 0) {
    return null
  }
  return { api_module: 'enabled', api_languages: langs }
}

function mapDocsSite(value: unknown): Record<string, unknown> | null {
  if (isOff(value, DOCS_SITE_OFF)) {
    return { docs_module: 'disabled' }
  }
  if (isSequence(value)) {
    return null
  }
  const framework = DOCS_SITE_FRAMEWORKS[normToken(value)]
  if (!framework) {
    return null
  }
  return { docs_module: 'enabled', docs_framework: framework }
}

function mapSaasStarterModule(value: unknown): Record<string, unknown> | null {
  const toggle = moduleToggle(value)
  if (toggle === null) {
    return null
  }
  return { saas_infra_module: toggle }
}

function mapSplit(
  value: unknown,
  moduleKey: string,
  providerKey: string,
  providers: ReadonlySet<string>,
): Record<string, unknown> | null {
  if (isOff(value, SPLIT_OFF)) {
    return { [moduleKey]: 'disabled' }
  }
  if (isSequence(value)) {
    return null
  }
  const provider = normToken(value)
  if (!providers.has(provider)) {
    return null
  }
  return { [moduleKey]: 'enabled', [providerKey]: provider }
}

function mapSaasAuth(value: unknown): Record<string, unknown> | null {
  return mapSplit(value, 'saas_auth_module', 'saas_auth_provider', SAAS_AUTH_PROVIDERS)
}

function mapSaasBilling(value: unknown): Record<string, unknown> | null {
  return mapSplit(
    value,
    'saas_billing_module',
    'saas_billing_provider',
    SAAS_BILLING_PROVIDERS,
  )
}

function mapIncludeAdmin(value: unknown): Record<string, unknown> | null {
  const flag = asBool(value)
  if (flag === null) {
    return null
  }
  return { saas_admin_dashboard: flag }
}

const DEST_MAPPERS: Record<string, (value: unknown) => Record<string, unknown> | null> = {
  api_tracks: mapApiTracks,
  api_language: mapApiLanguage,
  docs_site: mapDocsSite,
  mcp_language: mapMcpLanguage,
  saas_starter_module: mapSaasStarterModule,
  saas_auth: mapSaasAuth,
  saas_billing: mapSaasBilling,
  include_admin: mapIncludeAdmin,
}

function writeDests(
  out: Record<string, unknown>,
  dests: Record<string, unknown>,
): Record<string, unknown> {
  const after: Record<string, unknown> = {}
  for (const [key, value] of Object.entries(dests)) {
    if (!(key in out)) {
      out[key] = value
    }
    after[key] = out[key]
  }
  return after
}

/** Apply known removed-key remaps. Leave unmapped leftovers for reject. */
export function applyRemovedKeyRemaps(
  answers: Record<string, unknown>,
): RemapResult {
  const out = { ...answers }
  const ops: RemapOp[] = []
  for (const [old, spec] of Object.entries(ANSWER_KEY_REMAPS)) {
    if (!(old in out)) {
      continue
    }
    const before = out[old]
    const dests = DEST_MAPPERS[old](before)
    if (dests === null) {
      continue
    }
    const after = writeDests(out, dests)
    delete out[old]
    ops.push({
      old,
      new_keys: spec.new_keys,
      action: spec.action,
      before,
      after,
    })
  }
  return { answers: out, ops }
}

/** Plan name for the TS remap twin (same contract as `applyRemovedKeyRemaps`). */
export function remapRemovedAnswerKeys(
  answers: Record<string, unknown>,
): RemapResult {
  return applyRemovedKeyRemaps(answers)
}

export function findRemovedAnswerKeys(
  config: Record<string, unknown>,
): string[] {
  return Object.keys(config)
    .filter((key) => key in REMOVED_ANSWER_KEYS)
    .sort()
}

export function formatRemovedAnswerKeyErrors(
  config: Record<string, unknown>,
): string[] {
  return findRemovedAnswerKeys(config).map(
    (key) => `${key}: removed answer key; use ${REMOVED_ANSWER_KEYS[key]}`,
  )
}

function formatPreviewValue(value: unknown): string {
  if (Array.isArray(value)) {
    return `[${value.join(', ')}]`
  }
  if (typeof value === 'string') {
    return value
  }
  if (value === undefined) {
    return 'undefined'
  }
  return JSON.stringify(value)
}

export function formatRemapPreview(ops: RemapOp[]): string[] {
  return ops.map((op) => {
    const dests = Object.entries(op.after ?? {})
      .map(([key, value]) => `${key}=${formatPreviewValue(value)}`)
      .join(', ')
    return `${op.old}: ${formatPreviewValue(op.before)} → ${dests} (${op.action})`
  })
}

/** Apply remaps, then fail-closed on leftover removed keys or unmapped values. */
export function applyThenRejectRemovedKeys(
  answers: Record<string, unknown>,
): RemapResult {
  const result = applyRemovedKeyRemaps(answers)
  const leftover = formatRemovedAnswerKeyErrors(result.answers)
  if (leftover.length > 0) {
    throw new RemovedAnswerKeyError(leftover)
  }
  return result
}

/** Hydration helper: remap known keys and drop leftovers (never throw). */
export function dropLeftoverRemovedKeys(
  answers: Record<string, unknown>,
): Record<string, unknown> {
  const { answers: remapped } = applyRemovedKeyRemaps(answers)
  const out = { ...remapped }
  for (const key of findRemovedAnswerKeys(out)) {
    delete out[key]
  }
  return out
}

File: src/riso/core/removed_answer_keys.py
"""Canonical removed Copier answer keys (package SSOT for CLI/wheel installs)."""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Any

REMOVED_ANSWER_KEYS: dict[str, str] = {
    "api_tracks": "`api_module` plus `api_languages`",
    "api_language": "`api_languages`",
    "docs_site": "`docs_module` plus `docs_framework`",
    "mcp_language": "`mcp_languages`",
    "saas_starter_module": "`saas_infra_module`",
    "saas_auth": "`saas_auth_module` plus `saas_auth_provider`",
    "saas_billing": "`saas_billing_module` plus `saas_billing_provider`",
    "include_admin": "`saas_admin_dashboard`",
    # Note: graphql_api_module / websocket_module remain as *derived* Jinja flags
    # produced by pre_gen normalize_api_feature_modules — not user prompts, but
    # still written into answers context for template excludes. Do not list them
    # as removed user keys or pre_gen self-fails after normalization.
}

_TOKEN_SPLIT = re.compile(r"[\s,+/|]+")

_API_LANGUAGES = frozenset({"python", "node", "rust", "go"})
_MCP_LANGUAGES = frozenset({"python", "typescript", "rust", "go"})
_MCP_ALIASES = {"node": "typescript", "js": "typescript"}
_API_TRACK_ALIASES = {
    "python": "python",
    "node": "node",
    "rust": "rust",
    "go": "go",
    "fastapi": "python",
    "fastify": "node",
    "actix": "rust",
}
_API_TRACKS_OFF = frozenset({"", "none", "disabled", "[]"})
_DOCS_SITE_OFF = frozenset({"none", "false", "disabled", "off"})
_DOCS_SITE_FRAMEWORKS = {
    "sphinx": "sphinx-shibuya",
    "sphinx-shibuya": "sphinx-shibuya",
    "docusaurus": "docusaurus",
    "fumadocs": "fumadocs",
}
_SPLIT_OFF = frozenset({"none", "disabled", "false", "off"})
_SAAS_AUTH_PROVIDERS = frozenset({"clerk", "authjs"})
_SAAS_BILLING_PROVIDERS = frozenset({"stripe", "paddle", "lemonsqueezy"})
_MODULE_ON = frozenset({"enabled", "true", "yes", "on", "1"})
_MODULE_OFF = frozenset({"disabled", "false", "no", "off", "0", "none"})
_BOOL_TRUE = frozenset({"true", "yes", "on", "1", "enabled"})
_BOOL_FALSE = frozenset({"false", "no", "off", "0", "disabled", "none", ""})


@dataclass(frozen=True)
class RemapOp:
    """One remap-table row, or one applied operation when before/after are set."""

    old: str
    new_keys: tuple[str, ...]
    action: str
    before: Any = None
    after: Any = None


@dataclass(frozen=True)
class RemapResult:
    """Outcome of applying known removed-key remaps."""

    answers: dict[str, Any]
    ops: tuple[RemapOp, ...]


ANSWER_KEY_REMAPS: dict[str, RemapOp] = {
    "api_tracks": RemapOp(
        old="api_tracks",
        new_keys=("api_module", "api_languages"),
        action="derive",
    ),
    "api_language": RemapOp(
        old="api_language",
        new_keys=("api_languages",),
        action="wrap-list",
    ),
    "docs_site": RemapOp(
        old="docs_site",
        new_keys=("docs_module", "docs_framework"),
        action="derive",
    ),
    "mcp_language": RemapOp(
        old="mcp_language",
        new_keys=("mcp_languages",),
        action="wrap-list",
    ),
    "saas_starter_module": RemapOp(
        old="saas_starter_module",
        new_keys=("saas_infra_module",),
        action="rename",
    ),
    "saas_auth": RemapOp(
        old="saas_auth",
        new_keys=("saas_auth_module", "saas_auth_provider"),
        action="split",
    ),
    "saas_billing": RemapOp(
        old="saas_billing",
        new_keys=("saas_billing_module", "saas_billing_provider"),
        action="split",
    ),
    "include_admin": RemapOp(
        old="include_admin",
        new_keys=("saas_admin_dashboard",),
        action="rename-bool",
    ),
}


def _norm_token(value: Any) -> str:
    return str(value).strip().lower()


def _is_sequence(value: Any) -> bool:
    return isinstance(value, (list, tuple, set, frozenset))


def _tokens(value: Any) -> list[str]:
    if value is None:
        return []
    if _is_sequence(value):
        return [_norm_token(item) for item in value if _norm_token(item)]
    text = str(value).strip()
    if not text:
        return []
    return [part for part in _TOKEN_SPLIT.split(text.lower()) if part]


def _wrap_list(
    value: Any,
    allowed: frozenset[str],
    aliases: Mapping[str, str] | None = None,
) -> list[str] | None:
    alias_map = dict(aliases or {})
    if _is_sequence(value):
        out: list[str] = []
        seen: set[str] = set()
        for item in value:
            raw = str(item).strip()
            if not raw:
                continue
            mapped = alias_map.get(raw.lower(), raw.lower())
            if mapped not in allowed:
                return None
            if mapped in seen:
                continue
            seen.add(mapped)
            out.append(mapped)
        return out
    if isinstance(value, str):
        raw = value.strip()
        if not raw:
            return None
        mapped = alias_map.get(raw.lower(), raw.lower())
        if mapped not in allowed:
            return None
        return [mapped]
    return None


def _module_toggle(value: Any) -> str | None:
    if isinstance(value, bool):
        return "enabled" if value else "disabled"
    if (
        isinstance(value, (int, float))
        and value in {0, 1}
        and not isinstance(value, bool)
    ):
        return "enabled" if value == 1 else "disabled"
    text = _norm_token(value)
    if text in _MODULE_ON:
        return "enabled"
    if text in _MODULE_OFF:
        return "disabled"
    return None


def _as_bool(value: Any) -> bool | None:
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)) and value in {0, 1}:
        return bool(value)
    if isinstance(value, str):
        text = value.strip().lower()
        if text in _BOOL_TRUE:
            return True
        if text in _BOOL_FALSE:
            return False
    return None


def _is_off(value: Any, off_tokens: frozenset[str]) -> bool:
    if value is None:
        return True
    if _is_sequence(value) and len(value) == 0:
        return True
    if isinstance(value, bool):
        return (not value) and "false" in off_tokens
    return _norm_token(value) in off_tokens


def _map_api_language(value: Any) -> dict[str, Any] | None:
    wrapped = _wrap_list(value, _API_LANGUAGES)
    if wrapped is None:
        return None
    return {"api_languages": wrapped}


def _map_mcp_language(value: Any) -> dict[str, Any] | None:
    wrapped = _wrap_list(value, _MCP_LANGUAGES, _MCP_ALIASES)
    if wrapped is None:
        return None
    return {"mcp_languages": wrapped}


def _map_api_tracks(value: Any) -> dict[str, Any] | None:
    if _is_off(value, _API_TRACKS_OFF):
        return {"api_module": "disabled"}
    langs: list[str] = []
    seen: set[str] = set()
    for token in _tokens(value):
        mapped = _API_TRACK_ALIASES.get(token)
        if mapped is None or mapped in seen:
            continue
        seen.add(mapped)
        langs.append(mapped)
    if not langs:
        return None
    return {"api_module": "enabled", "api_languages": langs}


def _map_docs_site(value: Any) -> dict[str, Any] | None:
    if _is_off(value, _DOCS_SITE_OFF):
        return {"docs_module": "disabled"}
    if _is_sequence(value):
        return None
    framework = _DOCS_SITE_FRAMEWORKS.get(_norm_token(value))
    if framework is None:
        return None
    return {"docs_module": "enabled", "docs_framework": framework}


def _map_saas_starter_module(value: Any) -> dict[str, Any] | None:
    toggle = _module_toggle(value)
    if toggle is None:
        return None
    return {"saas_infra_module": toggle}


def _map_split(
    value: Any,
    *,
    module_key: str,
    provider_key: str,
    providers: frozenset[str],
) -> dict[str, Any] | None:
    if _is_off(value, _SPLIT_OFF):
        return {module_key: "disabled"}
    if _is_sequence(value):
        return None
    provider = _norm_token(value)
    if provider not in providers:
        return None
    return {module_key: "enabled", provider_key: provider}


def _map_saas_auth(value: Any) -> dict[str, Any] | None:
    return _map_split(
        value,
        module_key="saas_auth_module",
        provider_key="saas_auth_provider",
        providers=_SAAS_AUTH_PROVIDERS,
    )


def _map_saas_billing(value: Any) -> dict[str, Any] | None:
    return _map_split(
        value,
        module_key="saas_billing_module",
        provider_key="saas_billing_provider",
        providers=_SAAS_BILLING_PROVIDERS,
    )


def _map_include_admin(value: Any) -> dict[str, Any] | None:
    flag = _as_bool(value)
    if flag is None:
        return None
    return {"saas_admin_dashboard": flag}


_DEST_MAPPERS = {
    "api_tracks": _map_api_tracks,
    "api_language": _map_api_language,
    "docs_site": _map_docs_site,
    "mcp_language": _map_mcp_language,
    "saas_starter_module": _map_saas_starter_module,
    "saas_auth": _map_saas_auth,
    "saas_billing": _map_saas_billing,
    "include_admin": _map_include_admin,
}


def _write_dests(out: dict[str, Any], dests: Mapping[str, Any]) -> dict[str, Any]:
    after: dict[str, Any] = {}
    for key, value in dests.items():
        if key not in out:
            out[key] = value
        after[key] = out[key]
    return after


def apply_removed_key_remaps(answers: Mapping[str, Any]) -> RemapResult:
    """Apply known removed-key remaps. Leave unmapped leftovers for reject."""
    out = dict(answers)
    ops: list[RemapOp] = []
    for old, spec in ANSWER_KEY_REMAPS.items():
        if old not in out:
            continue
        before = out[old]
        dests = _DEST_MAPPERS[old](before)
        if dests is None:
            continue
        after = _write_dests(out, dests)
        del out[old]
        ops.append(
            RemapOp(
                old=old,
                new_keys=spec.new_keys,
                action=spec.action,
                before=before,
                after=after,
            )
        )
    return RemapResult(answers=out, ops=tuple(ops))


__all__ = [
    "ANSWER_KEY_REMAPS",
    "REMOVED_ANSWER_KEYS",
    "RemapOp",
    "RemapResult",
    "apply_removed_key_remaps",
]
